### DISCUSSION TITLE
Lyrics integration?

### DISCUSSION THREAD

**Discussion by antoinevandenhurk** [music-assistant/support#1652] [Feature requests and ideas]
Dear music lovers,

The 'music assistant' keeps getting better and better. Thank you very much for that, I enjoy using this feature!
Wouldn't it be great to integrate the lyrics of the playing song into the MASS interface? In any case, I would certainly appreciate it!

Perhaps it could be done the way Robert did it. [https://github.com/robert-alfaro/genius-lyrics](url)

Have a nice day...

**Comment by OzGav** [comment]
Thanks for the suggestion. This won’t happen before the 2.0 release. After that we will see how many votes this feature gets…

**Reply by max246** [reply]
Any news?

**Reply by OzGav** [reply]
No. I don’t mind the idea but with 8 votes it is way down the list at the moment.

**Reply by max246** [reply]
> No. I don’t mind the idea but with 8 votes it is way down the list at the moment.

I am up for testing out something, I am not familar with this code base but I suppose Spotify might expose the endpoint.
I might thinker around with the project and see if there is some interest in reviewing my code

**Reply by edisun3003** [reply]
Roon as a method to choose chromecast enabled tvs as display for artwork and stnced lyrics. I want this feature. It even works on a web ui server-running-roon:9330

**Reply by OzGav** [reply]
It isn’t clear what you are asking for here but I don’t see we need anything from Roon when we have existing providers for lyrics.

**Comment by zkvsky** [comment]
Current releases of Jellyfin have lyrics support integrated, it would be nice to make use of that.

**Comment by Chris78c** [comment]
Sync Lyrics would be the best. I vote for this suggestion.

**Reply by antoinevandenhurk** [reply]
Good suggestion

**Reply by AnshulJ999** [reply]
+1 for this. Any method of having synced lyrics in HA would be so cool.

**Reply by fangming828** [reply]
+1 for this. Any method of having synced lyrics in HA would be so cool.

**Comment by OzGav** [comment]
Lyrics support for Tidal is coming in the next 2.6.0 betas. Support for other providers will follow thereafter.

**Reply by antoinevandenhurk** [reply]
That's great!

**Reply by wessamlauf** [reply]
Awesome! I am YouTube music user so I hope one day it'll arrive there as well !

**Reply by OzGav** [reply]
Actually it already has for items that are in your library and that have lyrics in the lrclib database. You just have to update the metadata.

**Reply by Vergil365** [reply]
> Lyrics support for Tidal is coming in the next 2.6.0 betas. Support for other providers will follow thereafter.

Is it planned to support synced lyrics for local song (like subsonic)?
Navidrome uses the TAG "Lyrics" in this format to show synced lyrics:

[00:00.00] 作曲 : Colin Brittain/Jacoby Shaddix/Jason Aalon/Nick Furlong/Spencer Nezey/Sueco/Tobin Esperance
[00:23.42] Bitch don't play with us,
[00:24.69] We far too dangerous
[00:26.08] End up with a noose around your neck
...

But unfortunately there is an issue that no one tries to fix, so it would be amazing to have this function in music assistant.

Or a service call to get synced lyrics from musixmatch no matter where the music from music assistant comes from would be a gamechanger.
This would make it possible to have synced lyrics even for radio stations.

What is your opinion?

**Reply by OzGav** [reply]
I have already addressed this in the post immediately above yours. If it’s in the library you can get lyrics within any constraints of the lrclib database.

**Reply by Vergil365** [reply]
> I have already addressed this in the post immediately above yours. If it’s in the library you can get lyrics within any constraints of the lrclib database.

Oh sorry i did not understand what it means.
So after some googling, lrclib contains the lyrics (like musixmatch) and i can use programs like foobar2000 to write this lyrics to my songs.

In the case of navidrome the lyrics are only shown if the MP3TAG is called "Lyrics", does this work for music assistant too or do we need another TAG?

**Reply by OzGav** [reply]
You can tag your music with lyrics but the plugin will pull the lyrics into the MA database for you. MA does understand the lyrics tag

**Reply by Vergil365** [reply]
> You can tag your music with lyrics but the plugin will pull the lyrics into the MA database for you. MA does understand the lyrics tag

Great, thank you very much, this is really a dream come true 😃

**Reply by Vergil365** [reply]
> You can tag your music with lyrics but the plugin will pull the lyrics into the MA database for you. MA does understand the lyrics tag

- - I installed the beta version 2.6.0 and added my music with the subsonic provider, but i dont see lyrics in music assistant.
Where should it be displayed?
In the player?
![image](https://github.com/user-attachments/assets/92214569-420f-4a05-9eb7-0ef56b20699f)

**Reply by OzGav** [reply]
https://www.music-assistant.io/music-providers/subsonic/ 

Lyrics not supported yet with that provider

**Reply by Vergil365** [reply]
> https://www.music-assistant.io/music-providers/subsonic/ 
> 
> 
> 
> Lyrics not supported yet with that provider 

Ok, i thought music assistant can read the "Lyrics" tag no matter what provider.
in this case i will be patient.
thank you for your work 😊

**Reply by ben-umble** [reply]
> https://www.music-assistant.io/music-providers/subsonic/
> 
> Lyrics not supported yet with that provider

Are there plans in place to support lyrics with this provider? Kind of a deal breaker, and I'm happy to PR if there's nothing in motion yet :)

To be clear, I'd like MA to read the `LYRICS` tag on tracks served from Navidrome specifically (but in theory any Subsonic provider). Symfonium on Android does this so it should be possible.

**Reply by edisun3003** [reply]
> > You can tag your music with lyrics but the plugin will pull the lyrics into the MA database for you. MA does understand the lyrics tag
> 
> - - I installed the beta version 2.6.0 and added my music with the subsonic provider, but i dont see lyrics in music assistant.
> Where should it be displayed?
> In the player?
> ![image](https://github.com/user-attachments/assets/92214569-420f-4a05-9eb7-0ef56b20699f)

If you click on the queue button it sometimes appears as a 3rd tab - Played - Queue - Lyrics like that

**Comment by antoinevandenhurk** [comment]
I love how the lyrics are integrated. Is this still a work in progress? Maybe an adjustable sync option would be a good addition. Some lyrics are not completely in sync depending on the (original) music file. Nevertheless, great work so far!

**Reply by OzGav** [reply]
Thanks. It is very difficult to get the lyrics perfectly in sync with the large array of player devices that MA supports. This also needs to be combined with player and network variances in latency. Couple that with the chance of a streaming player provider padding or cutting the start of a track. Improvments are being investigated but we are keeping the realities as to what is technicaly feasible in check. The adjustable sync option may not even work as currently MA is constantly recalculating the sync offset. If you had to keep doing that throughout a track you would be better off just scrolling the lyrics manually.

**Reply by sgzFalcon** [reply]
Maybe it's just a matter of waiting for the stable release but I'm running 2.6.0b14 using filesystem/local files as a provider and I have a .lrc file next to each .mp3 music file with the same name but lyrics don't seem to appear anywhere. Am I missing something? Thank you

**Reply by OzGav** [reply]
Please show the folder structure you are using.

**Reply by sgzFalcon** [reply]
This is my folder:
```
music  
├── A Great Big World  
│   └── When the Morning Comes  
│       ├── 1. All I Want Is Love.lrc  
│       ├── 1. All I Want Is Love.mp3  
│       ├── 12. Hold Each Other.lrc  
│       ├── 12. Hold Each Other.mp3  
│       └── cover.jpg  
├── Adelitas Way  
├── AFROJACK  
├── Against The Current  
└── ...
```

I'm using docker and I have mapped the volume to the music folder:
```
volumes:
      - ${USERDIR:-$HOME}/docker/music-assistant-server/data:/data/
      - ${USERDIR:-$HOME}/music:/music
```
In the settings of MA Provider I have path=/music

**Reply by OzGav** [reply]
When you say the lyrics don't appear anywhere where are you looking? Have you refreshed the metadata since the lyrics file was added?

**Reply by sgzFalcon** [reply]
I was expecting to see them when using the full screen player just like @Vergil365 [showed before](https://github.com/orgs/music-assistant/discussions/1652#discussioncomment-13278789). I've tried Sync Now and Reload in the provider and the tracks view without success. Also, the artist page with Refresh Metadata which actually added a small description and images. I went as far as resetting the whole DB: `MA SETTINGS > CORE > MUSIC CONTROLLER`

**Reply by OzGav** [reply]
Resetting the DB probably should have done it but is not necessary. You just need to go to the track and in the menu in the top right select UPDATE METADATA. So is the lyrics file identical (case and everything) to the mp3 filename? This should work if you are on the beta which you say you are

**Reply by sgzFalcon** [reply]
They are identical:
![image](https://github.com/user-attachments/assets/5d600ac9-95a3-44c6-8e9d-2b9bc7ab4b6b)
I checked the version again:
![image](https://github.com/user-attachments/assets/cc476277-593d-4fe5-a6a4-01a253064032)
Sorry for all this trouble. I'm sure is something very silly on my side but I ran out of ideas.

**Reply by OzGav** [reply]
@fmunkes Any ideas?

**Reply by fmunkes** [reply]
2 questions:
* What's the lrc file's encoding?
* Can you post the first couple lines of one of the lrc files?

**Reply by sgzFalcon** [reply]
Of course, here you have the first lines:
```
[00:00.38] Oh oh whoa oh oh
[00:03.27] Oh oh whoa oh oh
[00:05.49] Oh oh whoa oh oh
[00:09.98] Fear
```
Encoding is UTF-8. I got all the files from lrclib using [LRCGET](https://github.com/tranxuanthang/lrcget)

**Reply by sgzFalcon** [reply]
It was me indeed who didn't know where to look for them. In case there are more people like me out there:
1.  Select the player
2.  Select the queue
3.  On the top options click on Lyrics
![image](https://github.com/user-attachments/assets/5d82ea61-6790-47cc-9df0-7afe7ad01533)

Thank you @OzGav and @fmunkes

**Comment by iclanzan** [comment]
Are lyrics from Jellyfin supported yet?

**Reply by OzGav** [reply]
Next beta there is a change that might enable 5his. I haven’t got Jellyfin to try it.

**Comment by edisun3003** [comment]
For ex, in Roon I can stream the music to my amplifier via airplay or raat server and then for DISPLAYING ALONE i can select the chromecast enabled TV . Attaching the image
<img width="1170" height="1332" alt="image" src="https://github.com/user-attachments/assets/d4a85d62-034e-42d5-92ea-95be57f8d629" />

**Reply by edisun3003** [reply]
In this pic the music is streamed to my Marantz-MCR612 via airplay , for the display i selected my chromecast enabled Sony TV which displays artwork and time synced lyrics from lrc file or embedded lyrics

**Reply by edisun3003** [reply]
![image](https://github.com/user-attachments/assets/98e1b9a7-669d-419b-87df-c4184bc0a33c)

Attaching the display settings page of roon if u can get a better idea of what im saying.

**Reply by OzGav** [reply]
Ok this has little to do with adding lyrics support. You need to open your own feature request for secondary display support (as that’s what it sounds like to me)

**Comment by maxi1134** [comment]
I downloaded lyrics for all my tracks and can see them in the MA sub-sub-sub-menu.

Any way to get this in a HA card for easy display on tablets?

**Reply by OzGav** [reply]
Good question. The answer is not at the moment as that info isn't exposed to the HA action responses. I'll keep it in mind for future updates

**Reply by maxi1134** [reply]
> Good question. The answer is not at the moment as that info isn't exposed to the HA action responses. I'll keep it in mind for future updates

Alternatively, a way to access a MA lyrics page with a dedicated URL would work! THen we can Iframe that page; if that's easier to build than an integration with HA

**Reply by OzGav** [reply]
I'm not sure but have a look at the new API as you might find the info exposed there that you can use? https://beta.music-assistant.io/api/

**Comment by janusloo** [comment]
Hi I am using Tidal as my music provider and I can't see any lyrics for the song that is playing. I know I can activate it by going to the More Info menu, wait for it to fetch the song info, then refresh item, press Play again and the lyrics will be displayed in the UI. A bit cumbersome for the experience. Can't the lyrics be fetched on the fly when the song starts playing? It would be awesome if this is automatic and real time!

**Comment by floodwayprintco** [comment]
I am trying to figure it out with Spotify, as well. I'd LOVE to have lyrics scrolling on a Lovelace view! Throwing in a vote!

**Comment by AnshulJ999** [comment]
Hi, I apologize for the shameless self-promotion here, but I had followed this thread years ago and just got notified about it. 

I built my own solution to this: SyncLyrics - https://github.com/AnshulJ999/SyncLyrics ([Reddit thread for reference](https://www.reddit.com/r/homeassistant/comments/1q7elpv/synclyrics_realtime_synced_lyrics_for_your_ha/))

It can be installed as an Home Assistant addon (multiple other methods too) and has full support for Music Assistant as well as Spotify and a few other providers. Free and open-source of course. 

Might be particularly of interest to the last 2 comments I see here from @floodwayprintco and @janusloo as it's fully automatic, real-time, and can be embedded as an iFrame in Lovelace. 

I hope it might come in handy for some of the people here. Cheers :D

**Reply by frostworx** [reply]
I can confirm that SyncLyrics is excellent :)

**Reply by edisun3003** [reply]
Is it possible to link it with music-assistant-server running as a docker service

**Reply by AnshulJ999** [reply]
Hi @edisun3003 , I haven't tried that specific scenario but I think it should work. You just need to put a Music Assistant Server URL and an access token, and it should work. But if it doesn't, please raise an issue on the repo with some logs and I will look into it. Thanks!

**Reply by maxi1134** [reply]
Was this vibe-coded? Not judging, just avoiding those projects.

**Reply by AnshulJ999** [reply]
@maxi1134 , It was AI assisted and I mentioned it in the README too, but considering I spent months of development work on it, I wouldn't call it 'vibe-coded'. Of course it's your choice whether you'd like to try it or not :)

**Reply by edisun3003** [reply]
Hey Anshul,
Yes it worked with Music-Assistant and Your service running as a docker container 

I used this
  synclyrics:
    image: anshulj99/synclyrics:latest
    container_name: synclyrics
    restart: unless-stopped
    network_mode: host
    # Add this line (replace 1000 with the result of 'id -u' if different)
    user: "1000:1000" 
    environment:
      - SYSTEM_MUSIC_ASSISTANT_SERVER_URL=http://192.168.0.143:8095
      - SYSTEM_MUSIC_ASSISTANT_TOKEN=your_token_here
      - AUDIODB_API_KEY=523532
    volumes:
      - ${USERDIR:-$HOME}/docker/synclyrics/data:/data

I inserted this along with the music-assistant compose yaml file

The problem is the the secret key and the Server address we input in the compose are not showing inside the UI
In the UI settings, i had to enter these 2 values inside the Music Assistant settings in the dashboard 

The Synced Lyrics appear really super. I love it
Mainly I love how the art work is displayed on the screen even when lyrics are not found. 

Can you make it to recongnize embedded lyrics too for downloaded music or lrc files stored along with music im the same folder in the same name?

**Comment by edisun3003** [comment]
Hey Anshul,
Yes it worked with Music-Assistant and Your service running as a docker container

I used this
  synclyrics:
    image: anshulj99/synclyrics:latest
    container_name: synclyrics
    restart: unless-stopped
    network_mode: host
    # Add this line (replace 1000 with the result of 'id -u' if different)
    user: "1000:1000"
    environment:
      - SYSTEM_MUSIC_ASSISTANT_SERVER_URL=http://192.168.0.143:8095
      - SYSTEM_MUSIC_ASSISTANT_TOKEN=your_token_here
      - AUDIODB_API_KEY=523532
    volumes:
      - ${USERDIR:-$HOME}/docker/synclyrics/data:/data

I inserted thsi Along with the music -assistant compose yaml file

The problem is the the secret key and the Server address we input in the compose are not showing inside the UI
In the UI settings, i had to enter these 2 values inside the Music Assistant settings in the dashboard

The Lyrics appear super. I love it



Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by edisun3003** [comment]
Mainly I love how the art work is displayed on the screen even when lyrics are not found.

Can you make to recongnize embedded lyrics too for downloaded music or lrc files stored along with music im the same folder in the same name?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by AnshulJ999** [reply]
Hey @edisun3003 ! 

Thanks so much for the feedback! 

So about your previous message: if the Music Assistant URL and Token have been entered as ENV variables, they won't show up in the Settings UI. It'll show them as blank, but don't worry, the ENV variables take precedence. I should probably make this clearer in the UI somehow. And yeah ENV is preferable than storing secrets in Settings (thats a plain JSON file). 

As for embedded lyrics / LRC format; I don't really plan on adding support for this as it'd add a lot of complexity. The app already searches through 4-5 providers for lyrics and stores them in the database as simple JSON files. Generally you just need to play through your library once, and next time, it will load from database to be faster. Supporting multiple formats means I'd have to add LRC, ID3, etc., and that's incompatible with the multi-provider setup for now. 

In my tests, if SyncLyrics was unavailable to find lyrics for a song, they most likely do not exist. And I found it works for >95% of songs, even niche songs. But if you find a case where lyrics for a song do exist but my app did not find it, please create an issue on my repo and I will look into it.

**Comment by Schnurzel700** [comment]
Any update on support for Lyrics from Subsonic (Navidrome)?

**Reply by khers** [reply]
https://github.com/music-assistant/server/pull/3424

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
